### PR TITLE
[extractor/odnoklassniki] Boosty.to support

### PR DESCRIPTION
**Pull request by megapro17**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

Fix for 1440p and 2160p quality sorting

Example: https://ok.ru/video/313203755634

Add boosty.to support, fix https://github.com/yt-dlp/yt-dlp/issues/4212

Thanks for help @Lesmiscore 

Example:
https://boosty.to/vitaly/posts/acd4db50-18a8-40a0-a0bc-3bdee2022dbe
https://ok.ru/videoembed/3696046058009

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by pukkandan** [comment]
Add test

**Comment by megapro17** [comment]
well i need to remove that test or change regex?

**Comment by pukkandan** [comment]
embed tests go under `_WEBPAGE_TESTS`, not `_TESTS`

**Comment by Murtaught** [comment]
I've cloned @megapro17 's repo and it seems to work great.
I've managed to download paid videos from Boosty by asking **yt-dlp** to use cookies from my browser like so:

```
./yt-dlp.sh --cookies-from-browser firefox https://boosty.to/CREATOR/posts/aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee?share=post_link
```

Click on "share" button under the video to get link.
Thank you very much! 👍🏻

File: yt_dlp/extractor/odnoklassniki.py [hunk 1]
```diff
@@ -174,16 +242,26 @@ def _extract_desktop(self, url):
...
+                'Referer': smuggled['referrer'],
+            } if smuggled.get('referrer') else {})
 
         error = self._search_regex(
             r'[^>]+class="vp_video_stub_txt"[^>]*>([^<]+)<',
             webpage, 'error', default=None)
-        if error:
+        # Direct link from boosty
+        if error == "The author of this video has not been found or is blocked" and not smuggled.get('referrer') and mode == 'videoembed':
+            return self._extract_desktop(smuggle_url(url, {'referrer': "https://boosty.to"}))
```
Comment (pukkandan) [inline yt_dlp/extractor/odnoklassniki.py:263]: line too long
```suggestion
        if (error == 'The author of this video has not been found or is blocked'
                and not smuggled.get('referrer') and mode == 'videoembed'):
            return self._extract_desktop(smuggle_url(url, {'referrer': 'https://boosty.to'}))
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/odnoklassniki.py [hunk 2]
```diff
@@ -30,135 +33,200 @@ class OdnoklassnikiIE(InfoExtractor):
...
-            'thumbnail': 'https://coub-anubis-a.akamaized.net/coub_storage/coub/simple/cw_image/c5ac87553bd/608e806a1239c210ab692/1545580913_00026.jpg',
-            'title': 'Народная забава',
-            'uploader': 'Nevata',
-            'upload_date': '20181223',
-            'age_limit': 0,
-            'uploader_id': 'nevata.s',
-            'like_count': int,
-            'duration': 8.08,
-            'repost_count': int,
+    _TESTS = [
```
[the remark this answers is not in the document]
Comment (megapro17) [reply yt_dlp/extractor/odnoklassniki.py:36]: can u please tell me how? i just ran flake8

File: yt_dlp/extractor/odnoklassniki.py [hunk 3]
```diff
@@ -214,7 +292,7 @@ def _extract_desktop(self, url):
         movie = metadata['movie']
 
         # Some embedded videos may not contain title in movie dict (e.g.
-        # http://ok.ru/video/62036049272859-0) thus we allow missing title
+        # https://ok.ru/video/62036049272859-0) thus we allow missing title
```
[the remark this answers is not in the document]
Comment (megapro17) [reply yt_dlp/extractor/odnoklassniki.py:295]: is there any reason to put http, it redirects to https anyway

File: yt_dlp/utils.py [hunk 4]
```diff
@@ -5923,6 +5923,15 @@ def orderedSet_from_options(options, alias_dict, *, use_regex=False, start=None)
     return orderedSet(requested)
 
 
+def scheme_host_only(url, origin=False):
+    parts = urllib.parse.urlsplit(url)
+    if parts.scheme in ('data',):
+        return None
+    url = urllib.parse.urlunsplit(parts._replace(
+        path='' if origin else '/', query='', fragment=''))
+    return url
```
[the remark this answers is not in the document]
Comment (megapro17) [reply yt_dlp/utils.py:5932]: @Lesmiscore please help

[the remark this answers is not in the document]
Comment (bashonly) [reply yt_dlp/utils.py:5932]: `base_url(url)` returns the same as `scheme_host_only(url)`

and `base_url(url).rstrip('/')` == `scheme_host_only(url, origin=True)`

File: yt_dlp/extractor/odnoklassniki.py [hunk 5]
```diff
@@ -64,12 +67,13 @@ class OdnoklassnikiIE(InfoExtractor):
         },
     }, {
         # metadata in JSON
-        'url': 'http://ok.ru/video/20079905452',
-        'md5': '0b62089b479e06681abaaca9d204f152',
+        'url': 'https://ok.ru/video/20079905452',
+        'md5': '5d2b64756e2af296e3b383a0bc02a6aa',
```
Comment (pukkandan) [inline yt_dlp/extractor/odnoklassniki.py:71]: Do u know why all the md5s changed?

File: yt_dlp/extractor/odnoklassniki.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/odnoklassniki.py]: fixed. flake8 doesnt change files. Just disable whatever auto-formatter u r using

File: yt_dlp/extractor/odnoklassniki.py [hunk 6]
```diff
@@ -30,135 +33,200 @@ class OdnoklassnikiIE(InfoExtractor):
...
+                'uploader_id': '571368965883',
+                'like_count': 0,
+                'duration': 10444,
+            },
+        }
+    ]
+
+    @classmethod
+    def _extract_embed_urls(cls, url, webpage):
+        yield from (smuggle_url(x, {'referrer': scheme_host_only(url)}) for x in super()._extract_embed_urls(url, webpage))
```
[the remark this answers is not in the document]
Comment (megapro17) [reply yt_dlp/extractor/odnoklassniki.py]: It seems that it works without entire method, plain url and embedded

File: yt_dlp/extractor/odnoklassniki.py (hunk 5 above)
Comment (megapro17) [reply yt_dlp/extractor/odnoklassniki.py:71]: Yes, they're all correct

File: yt_dlp/extractor/odnoklassniki.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (megapro17) [reply yt_dlp/extractor/odnoklassniki.py]:
```
-    @classmethod
-    def _extract_embed_urls(cls, url, webpage):
-        for x in super()._extract_embed_urls(url, webpage):
-            yield smuggle_url(x, {'referrer': base_url(url)})
-
```
I've removed this method, and nothing changed, it still can extract. What it does?

[the remark this answers is not in the document]
Comment (Lesmiscore) [reply yt_dlp/extractor/odnoklassniki.py]: The method is used by GenericIE to find embeds in an another sites like blogs

[the remark this answers is not in the document]
Comment (megapro17) [reply yt_dlp/extractor/odnoklassniki.py]: Isn't this site is not meant to be embedded except boosty page?

> doesnt referer=url work?

How to test this thing?

[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/odnoklassniki.py]: I'm confused. You are asking us what code you wrote is doing??

[the remark this answers is not in the document]
Comment (megapro17) [reply yt_dlp/extractor/odnoklassniki.py]: I didn't wrote that part :)

**Pull request merged**